oeps/oep-0037-bp-test-data.rst

```diff
@@ -0,0 +1,180 @@
+
+.. code-block:: python
+
+    def account_retirement_example(data):
+        data.test_user = UserFactory.create()
+        data.test_superuser = SuperuserFactory.create()
+        data.original_email = data.test_user.email
+
+        # Misc. setup
+        PendingEmailChangeFactory.create(user=data.test_user)
+        UserOrgTagFactory.create(user=data.test_user, key='foo', value='bar')
+        UserOrgTagFactory.create(user=data.test_user, key='cat', value='dog')
+
+        CourseEnrollmentAllowedFactory.create(email=data.original_email)
+
+(JavaScript tests seeking to stub out data returned from a server can use similar functions utilizing a package like `rosie`_ or `factory-girl`_.)
+
+Setting attributes of the provided argument like this allows the test object or class to be provided when used in a unit test (so assertions can be written naturally), and something like a `SimpleNamespace`_ object can be provided when run from a developer console so the generated data can be quickly located and inspected.  When needed, course data should be loaded from an `OLX`_ archive via ``call_command()`` using the existing management commands for course import.
+
+Only the attributes needed to perform the intended task should be specified; the factory class should take care of setting reasonable defaults for the rest.  The `Faker`_ package can be used by factories to generate reasonable-looking values for many common types of data.  By auto-populating many of the fields this way, when a model changes often only the factory will need to be updated rather than every data set using that model.
+
+The function can take one or more positional arguments if useful for using it in different contexts.  It could for example create records linked to a provided existing user, or populate them based on information loaded from a YAML file found at the provided path.  Keyword-only arguments should usually be avoided, as they are more difficult to use from the kind of general-purpose management commands and ``make`` targets described below.
+
+.. _factory_boy: https://factoryboy.readthedocs.io/en/latest/index.html
+.. _rosie: https://www.npmjs.com/package/rosie
+.. _factory-girl: https://www.npmjs.com/package/factory-girl
+.. _SimpleNamespace: https://docs.python.org/3/library/types.html#types.SimpleNamespace
+.. _OLX: https://edx.readthedocs.io/projects/edx-open-learning-xml/en/latest/index.html
+.. _Faker: https://faker.readthedocs.io/en/stable/
+
+Loading Data
+------------
+
+Each IDA should provide a ``load_test_data`` management command which takes dotted paths to one or more test data generation functions as arguments.  For example, the command to load the example data above into the LMS might look something like this:
```

Comment:
Reading this OEP a bit late.  This may be a silly question, but if we do decide to use a dump file to speed up local development environments (such as for edx-platform), should `load_test_data` be primarily used to generate the data that is then dumped to a file and would there be another make target or place for this dump to be consumed?  In the current `enterprise-catalog` provisioning script for LMS we run `docker-compose exec -T mysql mysql edxapp < dump.sql` directly rather than using a make target, but I could see us not wanting to do that.
https://github.com/edx/enterprise-catalog/blob/master/decentralized_devstack/provision-lms.sh#L15